src/metabase/api/collection.clj

```diff
@@ -46,19 +46,52 @@
 
 (declare root-collection)
 
+(defn- location-from-collection-id-clause
+  "Clause to restrict which collections are being selected based off collection-id. If collection-id is nil,
+   then restrict to the children and the grandchildren of the root collection. If collection-id is an an integer,
+   then restrict to that collection's parents and children."
+  [collection-id]
+  (if collection-id
+    [:and
+     [:like :location (str "%/" collection-id "/%")]
+     [:not [:like :location (str "%/" collection-id "/%/%/%")]]]
+    [:not [:like :location "/%/%/"]]))
+
 (defn- remove-other-users-personal-collections
   [user-id collections]
-  (let [personal-ids (into #{} (comp (filter :personal_owner_id)
-                                     (remove (comp #{user-id} :personal_owner_id))
-                                     (map :id))
-                           collections)
-        prefixes     (into #{} (map (fn [id] (format "/%d/" id))) personal-ids)
-        personal?    (fn [{^String location :location id :id}]
-                       (or (personal-ids id)
-                           (prefixes (re-find #"^/\d+/" location))))]
-    (if (seq prefixes)
-      (remove personal? collections)
-      collections)))
+  (let [personal-ids        (t2/select-fn-set :id :model/Collection
+                                              {:where
+                                               [:and [:!= :personal_owner_id nil] [:!= :personal_owner_id user-id]]})
+        personal-descendant (comp personal-ids
```

Comment:
nit: it's a little more readable when predicate functions end in `?`, I think ([style guide](https://guide.clojure.style/#naming-predicates))
```suggestion
        personal-descendant? (comp personal-ids
```